contracts/token/ERC721/ERC721.sol

```diff
@@ -233,28 +212,45 @@ abstract contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Er
     }
 
     /**
-     * @dev Safely mints `tokenId` and transfers it to `to`.
+     * @dev Transfers `tokenId` from its current owner to `to`, or alternatively mints (or burns) if the current owner
+     * (or `to`) is the zero address.
      *
-     * Requirements:
-     *
-     * - `tokenId` must not exist.
-     * - If `to` refers to a smart contract, it must implement {IERC721Receiver-onERC721Received}, which is called upon a safe transfer.
+     * The `operatorCheck` argument is optional. If the value passed is non 0, then this function will check that
+     * `operatorCheck` is either the owner of the token, or approved to operate on the token (by the owner).
      *
      * Emits a {Transfer} event.
      */
-    function _safeMint(address to, uint256 tokenId) internal virtual {
-        _safeMint(to, tokenId, "");
-    }
+    function _update(
+        address to,
+        uint256 tokenId,
+        address operatorCheck
+    ) internal virtual returns (address) {
+        address from = _ownerOf(tokenId);
 
-    /**
-     * @dev Same as {xref-ERC721-_safeMint-address-uint256-}[`_safeMint`], with an additional `data` parameter which is
-     * forwarded in {IERC721Receiver-onERC721Received} to contract recipients.
-     */
-    function _safeMint(address to, uint256 tokenId, bytes memory data) internal virtual {
-        _mint(to, tokenId);
-        if (!_checkOnERC721Received(address(0), to, tokenId, data)) {
-            revert ERC721InvalidReceiver(to);
+        // Perform (optional) operator check
+        if (operatorCheck != address(0) && from != operatorCheck && !isApprovedForAll(from, operatorCheck) && getApproved(tokenId) != operatorCheck) {
+            revert ERC721InsufficientApproval(operatorCheck, tokenId);
```

Comment:
There is still `_isApprovedOrOwner` in the contract, which is overrideable, but it's not used here, so we need to do something about that function.

In https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4419 I had refactored that function in this way:

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/cdacda12173d540137d263ff3779985b0519ae9a/contracts/token/ERC721/ERC721.sol#L229-L246

I believe we should do the same here, without the `actualOwner` part.

Reply:
I did that, let me know if it works for you